Add type parameters for SHTC1 / SHTC3

dbrgn opened the pull request:
Fixes #7.

dbrgn committed:
Add type parameters for SHTC1 / SHTC3

rnestler commented on examples/linux.rs:
```diff
@@ -1,12 +1,12 @@
     let dev = I2cdev::new("/dev/i2c-1").unwrap();
     let address = 0x70; // SHTC3
-    let mut sht = ShtCx::new(dev, address, Delay);
+    let mut sht = shtcx::shtc3(dev, address, Delay);
```
Should the I2C address be hardcoded for the sensors? There would still be the generic variant, right?

dbrgn replied:
Hm, we could hardcode them, true. As long as we have the generic variant as an escape hatch.

dbrgn replied:
Done :)

rnestler commented on src/lib.rs:
```diff
@@ -225,9 +227,28 @@ impl Command {
+/// Type parameter: SHTC1.
+pub struct ShtC1;
+/// Type parameter: SHTC3.
+pub struct ShtC3;
```
You could also add SHTW2 and SHTW2_alt.

dbrgn replied:
If the implementation for the SHTC3 and the SHTW2 are identical, we should probably not create multiple variants, right?

dbrgn replied:
Covered by `ShtBasic` and a `shtw2` factory function.

rnestler commented on src/lib.rs:
```diff
@@ -225,9 +227,28 @@ impl Command {
+    /// The chosen target sensor.
+    sensor: PhantomData<S>,
     /// The concrete I²C device implementation.
     i2c: I2C,
```
Not related to this PR, but is it smart to take ownership of the I2C bus? I mean since it is a bus it should be shared right?

rnestler replied:
So the following does not work:
```diff
diff --git a/examples/linux.rs b/examples/linux.rs
index 5d0a25e..9156779 100644
--- a/examples/linux.rs
+++ b/examples/linux.rs
@@ -8,6 +8,9 @@ fn main() {
     let address = 0x70; // SHTC3
     let mut sht = shtcx::shtc3(dev, address, Delay);
 
+    let address = 0x71; // foo
+    let mut sht2 = shtcx::shtc3(dev, address, Delay);
+

```

dbrgn replied:
You can always `destroy()` the driver to get back the instance. But yes, it would be great if there were a better alternative. Since the `i2c` instance is borrowed mutably, it's not that easy though.

Wasn't there some kind of crate for sharing a bus?

dbrgn replied:
https://crates.io/crates/shared-bus

dbrgn replied:
Oh cool, it seems that no changes need to be done to the driver crates to support `shared-bus` :slightly_smiling_face: We could add an example for multiple sensors.

rnestler reviewed (commented):
Looks nice and simple enough :slightly_smiling_face:  I'm a tiny bit worried about code size explosion due to monomporphisation (does different code get generated for zero sized types?). But it should seldomly matter since one typically uses just one sensor.

dbrgn commented:
> I'm a tiny bit worried about code size explosion due to monomporphisation (does different code get generated for zero sized types?).

Probably, the alternative would be compile flags, but you didn't like those :stuck_out_tongue: 

> But it should seldomly matter since one typically uses just one sensor.

Yep. And the idea was to provide a generic variant with some tolerance in timing that should work for all sensors (in case you don't know in advance which one you'll have).

dbrgn committed:
Hardcode I²C addresses

dbrgn committed:
Restructure tests into submodules

dbrgn committed:
Add factory function for SHTW2

dbrgn committed:
Move wrapper types into separate module

dbrgn committed:
Move CRC function into separate module

dbrgn committed:
Only impl sleep/wakeup for SHTC3

dbrgn commented:
@rnestler after a lot of different trait based approaches I switched to a macro based approach. For example, the `sleep` and `wakeup` commands are implemented for the `ShtC3` and the `ShtGeneric` using a macro. To make the docs easier to read, I moved them into a separate trait called `LowPower`.

Now the next question would be how to design the API for measurements. Right now there are three methods available on all sensors:

```rust
pub fn measure(&mut self, mode: PowerMode) -> Result<Measurement, Error<E>>;
pub fn measure_temperature(&mut self, mode: PowerMode) -> Result<Temperature, Error<E>>;
pub fn measure_humidity(&mut self, mode: PowerMode) -> Result<Humidity, Error<E>>;
```

The problem is that they all require a `PowerMode` which doesn't make much sense on the SHTC1.

---

One idea would be to only use normal mode in the normal methods (available on all sensors), and to move the low power measurements into the `LowPower` trait:

```rust
pub fn measure(&mut self) -> Result<Measurement, Error<E>>;
pub fn measure_temperature(&mut self) -> Result<Temperature, Error<E>>;
pub fn measure_humidity(&mut self) -> Result<Humidity, Error<E>>;

trait LowPower {
    pub fn measure_lp(&mut self) -> Result<Measurement, Error<E>>;
    pub fn measure_temperature_lp(&mut self) -> Result<Temperature, Error<E>>;
    pub fn measure_humidity_lp(&mut self) -> Result<Humidity, Error<E>>;
}
```

This would be OK, but we'd lose the ability to parametrize the `PowerMode` of the measurement.

---

Another approach would be to generate different measurement commands for the different sensors:

```rust
impl ShtCx<ShtC1> {
    pub fn measure(&mut self) -> Result<Measurement, Error<E>>;
    pub fn measure_temperature(&mut self) -> Result<Temperature, Error<E>>;
    pub fn measure_humidity(&mut self) -> Result<Humidity, Error<E>>;
}

impl ShtCx<ShtC3> {
    pub fn measure(&mut self, mode: PowerMode) -> Result<Measurement, Error<E>>;
    pub fn measure_temperature(&mut self, mode: PowerMode) -> Result<Temperature, Error<E>>;
    pub fn measure_humidity(&mut self, mode: PowerMode) -> Result<Humidity, Error<E>>;
}

impl ShtCx<ShtGeneric> {
    pub fn measure(&mut self, mode: PowerMode) -> Result<Measurement, Error<E>>;
    pub fn measure_temperature(&mut self, mode: PowerMode) -> Result<Temperature, Error<E>>;
    pub fn measure_humidity(&mut self, mode: PowerMode) -> Result<Humidity, Error<E>>;
}
```

This would be more ergonomic (the `PowerMode` is simply not part of the API for the SHTC1), but the downside is that the docs get quite ugly, because every measure method is listed n times. (Note that the impls could be generated with a macro, so that would be fine.)

---

As a third option, we could choose the first option with parametrized methods:

```rust
pub fn measure(&mut self) -> Result<Measurement, Error<E>>;
pub fn measure_temperature(&mut self) -> Result<Temperature, Error<E>>;
pub fn measure_humidity(&mut self) -> Result<Humidity, Error<E>>;

trait LowPower {
    pub fn measure_with_mode(&mut self, mode: PowerMode) -> Result<Measurement, Error<E>>;
    pub fn measure_temperature_with_mode(&mut self, mode: PowerMode) -> Result<Temperature, Error<E>>;
    pub fn measure_humidity_with_mode(&mut self, mode: PowerMode) -> Result<Humidity, Error<E>>;
}
```

The method names look ugly though. (Unfortunately we cannot call them `measure` as well without bad consequences for ergonomics, since there's no overloading in Rust.)

---

Any ideas?

dbrgn committed:
Replace sensor models with sensor classes

dbrgn committed:
Remove low-power measurement capability for ShtBasic

dbrgn committed:
Add SHTC1 example

dbrgn committed:
Improve test coverage

rnestler commented:
> The problem is that they all require a PowerMode which doesn't make much sense on the SHTC1.

Why doesn't it make sense on the SHTC1? You are aware that it supports the same measure modes as the SHTC3, right? See the [datasheet](https://www.sensirion.com/fileadmin/user_upload/customers/sensirion/Dokumente/0_Datasheets/Humidity/Sensirion_Humidity_Sensors_SHTC1_Datasheet.pdf):
> These values can be reduced by using the low power measurement mode, see separate application note.

And the mentioned [application note](https://www.sensirion.com/fileadmin/user_upload/customers/sensirion/Dokumente/2_Humidity_Sensors/Sensirion_Humidity_Sensors_SHTC1_SHTW2_STSC1_Low_Power_Measurement_Mode.pdf)

dbrgn commented:
WTF, that datasheet is very confusing.

### SHTC3

![2020-02-05-132539_425x212_scrot](https://user-images.githubusercontent.com/105168/73841476-0bf2cd80-481b-11ea-9112-3c246d1bd33a.png)

![2020-02-05-132615_630x191_scrot](https://user-images.githubusercontent.com/105168/73841498-19a85300-481b-11ea-8d87-cb8a1ee3f7c2.png)

![2020-02-05-132636_874x60_scrot](https://user-images.githubusercontent.com/105168/73841530-26c54200-481b-11ea-9179-ab52a5428491.png)

### SHTC1

![2020-02-05-132714_409x157_scrot](https://user-images.githubusercontent.com/105168/73841606-3b093f00-481b-11ea-8d05-9926282a4aa6.png)

![2020-02-05-132731_871x139_scrot](https://user-images.githubusercontent.com/105168/73841670-596f3a80-481b-11ea-96a1-0c3a7c6f80c4.png)

![2020-02-05-132803_881x57_scrot](https://user-images.githubusercontent.com/105168/73841673-5bd19480-481b-11ea-9f55-7955bce23e60.png)

The *only* reference to low power mode is in a small footnote.

If that's really the case, then the whole refactoring doesn't make sense :confused: So in the end, the only difference in the protocol between SHTC1 and SHTC3 are missing sleep/wakeup commands and different max measurement timings?

rnestler commented:
> So in the end, the only difference in the protocol between SHTC1 and SHTC3 are missing sleep/wakeup commands and different max measurement timings?

Yes, as far as I'm aware.
